Content.Shared/Chat/V2/Moderation/SimpleCensor.cs

```diff
@@ -0,0 +1,340 @@
+                    censored.Insert(i+1, censored[i] != replaceWith ? ')' : replaceWith);
+                }
+            }
+
+            if (censored[i] != replaceWith)
+            {
+                censored[i] = originalInput[i];
+            }
+        }
+
+        // SO says this is fast...
+        return string.Concat(censored);
+    }
+
+    /// <summary>
+    /// Adds a l33tsp34k sanitization rule
+    /// </summary>
+    /// <returns>The censor for further configuration</returns>
+    public SimpleCensor WithSanitizeLeetSpeak()
+    {
+        _shouldSanitizeLeetspeak = true;
+
+        return BuildCharacterReplacements();
+    }
+
+    /// <summary>
+    /// Adds a l33tsp34k sanitization rule
+    /// </summary>
+    /// <returns>The censor for further configuration</returns>
+    public SimpleCensor WithSanitizeSpecialCharacters()
+    {
+        _shouldSanitizeSpecialCharacters = true;
+
+        return BuildCharacterReplacements();
+    }
+
+    public SimpleCensor WithRanges(UnicodeRange[] ranges)
+    {
+        _allowedUnicodeRanges = ranges;
+
+        return this;
+    }
+
+    public SimpleCensor WithCustomDictionary(string[] naughtyWords)
+    {
+        _censoredWords = naughtyWords;
+
+        return this;
+    }
+
+    public SimpleCensor WithFalsePositives(string[] falsePositives)
+    {
+        _falsePositives = falsePositives;
+
+        return this;
+    }
+
+    public SimpleCensor WithFalseNegatives(string[] falseNegatives)
+    {
+        _falseNegatives = falseNegatives;
+
+        return this;
+    }
+
+    public SimpleCensor WithLeetspeakReplacements(Dictionary<char, char> replacements)
+    {
+        _leetspeakReplacements = replacements.ToFrozenDictionary();
+
+        return this;
+    }
+
+    public SimpleCensor WithSpecialCharacterReplacements(Dictionary<char, char> replacements)
+    {
+        _leetspeakReplacements = replacements.ToFrozenDictionary();
+
+        return this;
+    }
+
+    private string CheckProfanity(string input, List<char> censored, string[] words, char replaceWith = '*')
+    {
+        foreach (var word in words)
+        {
+            var wordLength = word.Length;
+            var endOfFoundWord = 0;
+            var foundIndex = input.IndexOf(word, endOfFoundWord, StringComparison.OrdinalIgnoreCase);
+
+            while(foundIndex > -1)
+            {
+                endOfFoundWord = foundIndex + wordLength;
+
+                for (var i = 0; i < wordLength; i++)
+                {
+                    censored[foundIndex+i] = replaceWith;
+                }
+
+                foundIndex = input.IndexOf(word, endOfFoundWord, StringComparison.OrdinalIgnoreCase);
+            }
+        }
+
+        return input;
+    }
+
+    private char[] FindFalsePositives(List<char> chars, char replaceWith = '*')
+    {
+        var input = string.Concat(chars);
+
+        var output = Enumerable.Repeat(replaceWith, input.Length).ToArray();
+        var inputAsARr = input.ToArray();
+
```

Comment:
These allocations here look pretty bad.

Would it be better to just pass the input around as a stringbuilder or span or the likes.

Reply:
Probably. I would need to benchmark and benchmarking C# is not something I've done before.

I pinkie promise to write a benchmark test for this and tune this code path to get it broadly fast enough.

Reply:
At the moment the raw string is being passed around and re-allocated multiple times to manipulate it instead of some kind of list structure. Like there's roughly 8 or so allocations being done per message, with 2 stringbuilders, multiple lists, etc.